### DISCUSSION TITLE
How to  decrypte the data in Wireshark，OR，how to turn  off the DTLS

### DISCUSSION THREAD

**Discussion by lokenetwork** [paullouisageneau/libdatachannel#1043] [General]
![1700477049943](https://github.com/paullouisageneau/libdatachannel/assets/5117703/3229170b-6003-46e3-884f-20e1f1e42de8)

**Comment by chobie** [comment]
FYI: https://github.com/paullouisageneau/libdatachannel/issues/921

As one solution, when you don't set the MediaHandler, you can directly receive RTP packets. 
You can export that packet data into something that can be imported with tools like `text2pcap`.

**Reply by paullouisageneau** [reply]
I think @lokenetwork refers to SCTP data from data channels transported over DTLS, not RTP, which is encrypted as SRTP instead.

**Reply by chobie** [reply]
I see, I haven't looked into SCTP before, so I'll check it out next time. thanks.

**Comment by paullouisageneau** [comment]
DTLS is mandatory for WebRTC connection establishment, so there is no way to disable it from the application. You can however dump SCTP packets in `rtc::impl::SctpTransport::incoming()` and `rtc::impl::SctpTransport::outgoing()` in a format you can import in Wireshark for debugging purposes.

**Comment by voluntas** [comment]
You can add an SSLKEYLOGFILE output mechanism to libdatachannel and have Wireshark's DTLS read it.

- https://wiki.wireshark.org/TLS#using-the-pre-master-secret
- https://datatracker.ietf.org/doc/html/draft-thomson-tls-keylogfile-01

[![Image from Gyazo](https://i.gyazo.com/5ff824b0d86c1ea5fa2fcbeaa7447d2f.png)](https://gyazo.com/5ff824b0d86c1ea5fa2fcbeaa7447d2f)

**Reply by mglettig** [reply]
> You can add an SSLKEYLOGFILE output mechanism to libdatachannel and have Wireshark's DTLS read it.
> 
> * https://wiki.wireshark.org/TLS#using-the-pre-master-secret
> * https://datatracker.ietf.org/doc/html/draft-thomson-tls-keylogfile-01
> 
> [![Image from Gyazo](https://camo.githubusercontent.com/d63d825d1627e89478283709e958743eb68bc3d4c8346a6e2a251b11c9f1cfea/68747470733a2f2f692e6779617a6f2e636f6d2f35666638323462306438366331656135666132666362656161373434376432662e706e67)](https://gyazo.com/5ff824b0d86c1ea5fa2fcbeaa7447d2f)

In which format did you enter the Pre-Shared Key in Wireshark? Something like the following?
`CLIENT_RANDOM 40db03ea1f61877df4ca74ee94446e4e4acd65e448cb53da4417b81fa33cced1 af41f78680ec068300a452c4bb499bfa0776809aa6c68ade8e7a85710f2d6722e286af276e5a2bf5b9f344fcdd3ad9ca`

Unfortunately it seems like I'm not able to decrypt the traffic that I previously captured.

**Comment by chobie** [comment]
I'll make it a patch for now since it's for local development. It's convenient to easily check with Wireshark.

```
diff --git a/src/impl/dtlstransport.cpp b/src/impl/dtlstransport.cpp
index aa518241..0199df74 100644
--- a/src/impl/dtlstransport.cpp
+++ b/src/impl/dtlstransport.cpp
@@ -752,6 +752,7 @@ DtlsTransport::DtlsTransport(shared_ptr<IceTransport> lower, certificate_ptr cer
 		SSL_CTX_set_read_ahead(mCtx, 1);
 		SSL_CTX_set_quiet_shutdown(mCtx, 0); // send the close_notify alert
 		SSL_CTX_set_info_callback(mCtx, InfoCallback);
+		SSL_CTX_set_keylog_callback(mCtx, KeyLogCallback);
 
 		SSL_CTX_set_verify(mCtx, SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT,
 		                   CertificateCallback);
@@ -1051,6 +1052,18 @@ void DtlsTransport::InfoCallback(const SSL *ssl, int where, int ret) {
 	}
 }
 
+void DtlsTransport::KeyLogCallback(const SSL *ssl, const char* line) {
+    const char* log_file_path = std::getenv("SSLKEYLOGFILE");
+    if (log_file_path != nullptr) {
+        FILE *log_file = fopen(log_file_path, "a");
+        if (log_file != nullptr) {
+            fprintf(log_file, "%s\n", line);
+            fclose(log_file);
+        }
+    }
+}
+
+
 int DtlsTransport::BioMethodNew(BIO *bio) {
 	BIO_set_init(bio, 1);
 	BIO_set_data(bio, NULL);
diff --git a/src/impl/dtlstransport.hpp b/src/impl/dtlstransport.hpp
index d84b312e..f4011c4a 100644
--- a/src/impl/dtlstransport.hpp
+++ b/src/impl/dtlstransport.hpp
@@ -110,6 +110,7 @@ protected:
 
 	static int CertificateCallback(int preverify_ok, X509_STORE_CTX *ctx);
 	static void InfoCallback(const SSL *ssl, int where, int ret);
+	static void KeyLogCallback(const SSL *ssl, const char* line);
 
 	static int BioMethodNew(BIO *bio);
 	static int BioMethodFree(BIO *bio);
```

**Reply by lokenetwork** [reply]
It work, Thank you so much

**Comment by florandens** [comment]
Can someone help me I have the same problem and would also like to decrypt it I have already tried with the SSLKEYLOGFILE but that does not work for my DTLS in a STUN package, how can I solve this concretely so that I can still see the data.
